pkg/contractreader/observed.go

```diff
@@ -0,0 +1,159 @@
+	)
+	CrBatchSizes = promauto.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "contract_reader_batch_request_size",
+			Help: "The size of the batch request",
+		},
+		[]string{"chainID"},
+	)
+	CrErrors = promauto.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "contract_reader_errors",
+			Help: "The number of errors that occurred during the ChainReader's execution",
+		},
+		[]string{"chainID", "function", "contract"},
+	)
+)
+
+type Observed struct {
+	ContractReaderFacade
+	lggr    logger.Logger
+	chainID string
+
+	// Prometheus components for tracking metrics
+	directRequestsDurations *prometheus.HistogramVec
+	batchRequestsDurations  *prometheus.HistogramVec
+	batchSizes              *prometheus.CounterVec
+	errors                  *prometheus.CounterVec
+}
+
+func NewObserverReader(
+	cr ContractReaderFacade,
+	lggr logger.Logger,
+	chainID string,
+) *Observed {
+	return &Observed{
+		ContractReaderFacade:    cr,
+		lggr:                    lggr,
+		chainID:                 chainID,
+		directRequestsDurations: CrDirectRequestsDurations,
+		batchRequestsDurations:  CrBatchRequestsDurations,
+		batchSizes:              CrBatchSizes,
+		errors:                  CrErrors,
+	}
+}
+
+func (o *Observed) GetLatestValue(
+	ctx context.Context,
+	readIdentifier string,
+	confidenceLevel primitives.ConfidenceLevel,
+	params, returnVal any,
+) error {
+	start := time.Now()
+	err := o.ContractReaderFacade.GetLatestValue(ctx, readIdentifier, confidenceLevel, params, returnVal)
+	duration := time.Since(start)
+
+	contract, function := unpackReadIdentifier(readIdentifier)
+	o.directRequestsDurations.
+		WithLabelValues(o.chainID, contract, function).
+		Observe(float64(duration))
+
+	o.maybeTrackErrors(err, "GetLatestValue", contract+"-"+function)
+	o.lggr.Debugw("Observed GetLatestValue",
+		"chainID", o.chainID,
+		"contract", contract,
+		"function", function,
+		"readIdentifier", readIdentifier,
+		"millis", duration.Milliseconds(),
+	)
+	return err
+}
+
+func (o *Observed) BatchGetLatestValues(
+	ctx context.Context,
+	request types.BatchGetLatestValuesRequest,
+) (types.BatchGetLatestValuesResult, error) {
+	start := time.Now()
+	result, err := o.ContractReaderFacade.BatchGetLatestValues(ctx, request)
+	duration := time.Since(start)
+
+	o.batchRequestsDurations.
+		WithLabelValues(o.chainID).
+		Observe(float64(duration))
+
+	o.batchSizes.
+		WithLabelValues(o.chainID).
+		Add(float64(len(request)))
+
+	o.maybeTrackErrors(err, "BatchGetLatestValues", "")
+	o.lggr.Debugw("Observed BatchGetLatestValues",
+		"chainID", o.chainID,
+		"millis", duration.Milliseconds(),
+		"size", len(request),
+		"contracts", maps.Keys(request),
+	)
+	return result, err
+}
+
+func (o *Observed) maybeTrackErrors(err error, function string, contract string) {
+	if err == nil {
+		return
+	}
+	o.errors.
+		WithLabelValues(o.chainID, function, contract).
+		Inc()
+}
+
+func unpackReadIdentifier(readIdentifier string) (contract string, function string) {
```

Comment:
I guess this is safe as readIdentifier are never going to change?

Reply:
Probably it deserved a comment. Read identifier has the following format `address-contract-method`. It's unacceptable for the prom, because addresses change on every deployment and cardinality will explode. Therefore I needed to cut of the address part from the identifier